src/device/device_ledger.cpp

```diff
@@ -0,0 +1,2057 @@
+    #define INS_RESET                           0x02
+
+    #define INS_GET_KEY                         0x20
+    #define INS_PUT_KEY                         0x22
+    #define INS_GET_CHACHA8_PREKEY              0x24
+    #define INS_VERIFY_KEY                      0x26
+
+    #define INS_SECRET_KEY_TO_PUBLIC_KEY        0x30
+    #define INS_GEN_KEY_DERIVATION              0x32
+    #define INS_DERIVATION_TO_SCALAR            0x34
+    #define INS_DERIVE_PUBLIC_KEY               0x36
+    #define INS_DERIVE_SECRET_KEY               0x38
+    #define INS_GEN_KEY_IMAGE                   0x3A
+    #define INS_SECRET_KEY_ADD                  0x3C
+    #define INS_SECRET_KEY_SUB                  0x3E
+    #define INS_GENERATE_KEYPAIR                0x40
+    #define INS_SECRET_SCAL_MUL_KEY             0x42
+    #define INS_SECRET_SCAL_MUL_BASE            0x44
+
+    #define INS_DERIVE_SUBADDRESS_PUBLIC_KEY    0x46
+    #define INS_GET_SUBADDRESS                  0x48
+    #define INS_GET_SUBADDRESS_SPEND_PUBLIC_KEY 0x4A
+    #define INS_GET_SUBADDRESS_SECRET_KEY       0x4C
+
+    #define INS_OPEN_TX                         0x70
+    #define INS_SET_SIGNATURE_MODE              0x72
+    #define INS_GET_ADDITIONAL_KEY              0x74
+    #define INS_STEALTH                         0x76
+    #define INS_BLIND                           0x78
+    #define INS_UNBLIND                         0x7A
+    #define INS_VALIDATE                        0x7C
+    #define INS_MLSAG                           0x7E
+    #define INS_CLOSE_TX                        0x80
+
+
+    #define INS_GET_RESPONSE                    0xc0
+
+
+    void DeviceLedger::logCMD() {
+      if (apdu_verbose) {
+        char  strbuffer[1024];
+        sprintf(strbuffer, "%.02x %.02x %.02x %.02x %.02x ",
+          this->buffer_send[0],
+          this->buffer_send[1],
+          this->buffer_send[2],
+          this->buffer_send[3],
+          this->buffer_send[4]
+          );
+        buffer_to_str(strbuffer+strlen(strbuffer), (char*)(this->buffer_send+5), this->length_send-5);
+        MDEBUG( "CMD  :" << strbuffer);
+      }
+    }
+
+    void DeviceLedger::logRESP() {
+      if (apdu_verbose) {
+        char  strbuffer[1024];
+        sprintf(strbuffer, "%.02x%.02x ",
+          this->buffer_recv[this->length_recv-2],
+          this->buffer_recv[this->length_recv-1]
+          );
+        buffer_to_str(strbuffer+strlen(strbuffer), (char*)(this->buffer_recv), this->length_recv-2);
+        MDEBUG( "RESP :" << strbuffer);
+
+      }
+    }
+
+    /* -------------------------------------------------------------- */
+    DeviceLedger::DeviceLedger() : exiting(false) DEVICE_CONTROLE {
+      this->id = device_id++;
+      this->hCard   = 0;
+      this->hContext = 0;
+      this->reset_buffer();
+      if (this->id)
+        MDEBUG( "Device "<<this->id <<" Created");
+    }
+
+    DeviceLedger::DeviceLedger(const DeviceLedger &device): DeviceLedger() {
```

Comment:
Are the copy construtor and assignment ever used? I would just delete them, because its unlikely to be in use with this specific implementation.

Reply:
Rereading all, yes copy/assignment constructors are no more used. I removed all of them by marking them 
`=delete`